opl/opl_sdl.c

```diff
@@ -155,59 +138,19 @@ static void AdvanceTime(unsigned int nsamples)
     SDL_UnlockMutex(callback_queue_mutex);
 }
 
-#if SDL_BYTEORDER == SDL_LIL_ENDIAN
-  #define OPL_SHORT SDL_SwapLE16
-#else
-  #define OPL_SHORT SDL_SwapBE16
-#endif
-
-int opl_gain = 200;
-
-static void MixAudioFormat(Uint8 *dst, const Uint8 *src, Uint32 len)
-{
-    Sint16 src1, src2;
-    int dst_sample;
-
-    while (len--)
-    {
-        src1 = OPL_SHORT(*(Sint16 *)src);
-        src2 = OPL_SHORT(*(Sint16 *)dst);
-        src += 2;
-        dst_sample = src1 * opl_gain / 100 + src2;
-        if (dst_sample > SHRT_MAX)
-        {
-            dst_sample = SHRT_MAX;
-        }
-        else if (dst_sample < SHRT_MIN)
-        {
-            dst_sample = SHRT_MIN;
-        }
-        *(Sint16 *)dst = OPL_SHORT(dst_sample);
-        dst += 2;
-    }
-}
-
 // Call the OPL emulator code to fill the specified buffer.
 
 static void FillBuffer(uint8_t *buffer, unsigned int nsamples)
 {
-    // This seems like a reasonable assumption.  mix_buffer is
-    // 1 second long, which should always be much longer than the
-    // SDL mix buffer.
-    assert(nsamples < mixing_freq);
-
-    // OPL output is generated into temporary buffer and then mixed
-    // (to avoid overflows etc.)
-    OPL3_GenerateStream(&opl_chip, (Bit16s *) mix_buffer, nsamples);
-    MixAudioFormat(buffer, mix_buffer, nsamples * 2);
+    OPL3_GenerateStream(&opl_chip, (Bit16s *) buffer, nsamples);
```

Comment:
So, in order to still support `opl_gain` we would have to implement it somewhere in `OPL3_GenerateStream()`?

Reply:
I thought maybe we could implement this on the OpenAL side. I need to experiment with it.

Reply:
> Oh, wait, OAL offer to set GAIN for a channel, right?

Right. Although I have already restored a version of our previous approach.

Reply:
Yes, and it's fine. I just think it would be even better if we could hand this task over to OAL. 😉

Reply:
It works like that:
```c
alSourcef(player.source, AL_MAX_GAIN, 8.0f);
alSourcef(player.source, AL_GAIN, (ALfloat)opl_gain / 100.0f);
```
It seems `AL_MAX_GAIN` is `1.0f` by default.

Reply:
Is setting `AL_MAX_GAIN` to `8.0f` arbitrary, at least `> 1.0f`? We allow `opl_gain = 1000` in `src/m_misc.c`.

Anyway, this looks much cleaner to me!